pkg/services/tree/signature.go

```diff
@@ -200,3 +188,57 @@ func eACLRole(role acl.Role) eacl.Role {
 		panic(fmt.Sprintf("unexpected tree service ACL role: %s", role))
 	}
 }
+
+// checkEACL searches for the eACL rules that could be applied to the request
+// (a tuple of a signer key, his NeoFS role and a request operation).
+// It does not filter the request by the filters of the eACL table since tree
+// requests do not contain any "object" information that could be filtered and,
+// therefore, filtering leads to unexpected results.
+// The code was copied with the minor updates from the SDK repo:
+// https://github.com/nspcc-dev/neofs-sdk-go/blob/43a57d42dd50dc60465bfd3482f7f12bcfcf3411/eacl/validator.go#L28.
+func checkEACL(tb eacl.Table, signer []byte, role eacl.Role, op eacl.Operation) error {
+	for _, record := range tb.Records() {
+		// check type of operation
+		if record.Operation() != op {
+			continue
+		}
+
+		// check target
+		if !targetMatches(record, role, signer) {
+			continue
+		}
+
+		switch a := record.Action(); a {
+		case eacl.ActionAllow:
+			return nil
+		case eacl.ActionDeny:
+			return eACLErr(op, errors.New("DENY eACL rule"))
```

Comment:
Can we use private var for this `DENY eACL` error, to avoid allocations?

Reply:
sure, added vars for all errors in that file